% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/main.R
\name{clustify_lists}
\alias{clustify_lists}
\alias{clustify_lists.default}
\alias{clustify_lists.seurat}
\alias{clustify_lists.Seurat}
\title{Main function to compare scRNA-seq data to gene lists.}
\usage{
clustify_lists(input, ...)

\method{clustify_lists}{default}(input, per_cell = F,
  cluster_info = NULL, log_scale = T, cluster_col = "cluster",
  topn = 1000, cut = 0, marker, marker_inmatrix = T,
  genomen = 30000, metric = "hyper", output_high = TRUE,
  set_ident = T, ...)

\method{clustify_lists}{seurat}(input, per_cell = F,
  cluster_info = NULL, log_scale = T, cluster_col = "cluster",
  topn = 1000, cut = 0, marker, marker_inmatrix = T,
  genomen = 30000, metric = "hyper", output_high = TRUE,
  dr = "tsne", seurat_out = TRUE, threshold = 0, set_ident = T,
  ...)

\method{clustify_lists}{Seurat}(input, per_cell = F,
  cluster_info = NULL, log_scale = T, cluster_col = "cluster",
  topn = 1000, cut = 0, marker, marker_inmatrix = T,
  genomen = 30000, metric = "hyper", output_high = TRUE,
  dr = "tsne", seurat_out = TRUE, threshold = 0, set_ident = T,
  ...)
}
\arguments{
\item{input}{single-cell expression matrix or Seurat object}

\item{...}{passed to matrixize_markers}

\item{per_cell}{compare per cell or per cluster}

\item{cluster_info}{data.frame or vector containing cluster assignments per cell.
Order must match column order in supplied matrix. If a data.frame
provide the cluster_col parameters.}

\item{log_scale}{input data is natural log,
averaging will be done on unlogged data}

\item{cluster_col}{column in cluster_info with cluster number}

\item{topn}{number of top expressing genes to keep from input matrix}

\item{cut}{expression cut off from input matrix}

\item{marker}{matrix or dataframe of candidate genes for each cluster}

\item{marker_inmatrix}{whether markers genes are already in preprocessed matrix form}

\item{genomen}{number of genes in the genome}

\item{metric}{adjusted p-value for hypergeometric test, or jaccard index}

\item{output_high}{if true (by default to fit with rest of package),
-log10 transform p-value}

\item{set_ident}{whether the output seurat object has ident set to called type}

\item{dr}{stored dimension reduction}

\item{seurat_out}{output cor matrix or called seurat object}

\item{input}{seurat object}

\item{per_cell}{compare per cell or per cluster}

\item{cluster_info}{data.frame or vector containing cluster assignments per cell.
Order must match column order in supplied matrix. If a data.frame
provide the cluster_col parameters.}

\item{log_scale}{input data is natural log,
averaging will be done on unlogged data}

\item{cluster_col}{column in cluster_info with cluster number}

\item{topn}{number of top expressing genes to keep from input matrix}

\item{cut}{expression cut off from input matrix}

\item{marker}{matrix or dataframe of candidate genes for each cluster}

\item{marker_inmatrix}{whether markers genes are already in preprocessed matrix form}

\item{genomen}{number of genes in the genome}

\item{metric}{adjusted p-value for hypergeometric test, or jaccard index}

\item{output_high}{if true (by default to fit with rest of package),
-log10 transform p-value}

\item{...}{passed to matrixize_markers}

\item{input}{seurat object}

\item{per_cell}{compare per cell or per cluster}

\item{cluster_info}{data.frame or vector containing cluster assignments per cell.
Order must match column order in supplied matrix. If a data.frame
provide the cluster_col parameters.}

\item{log_scale}{input data is natural log,
averaging will be done on unlogged data}

\item{cluster_col}{column in cluster_info with cluster number}

\item{topn}{number of top expressing genes to keep from input matrix}

\item{cut}{expression cut off from input matrix}

\item{marker}{matrix or dataframe of candidate genes for each cluster}

\item{marker_inmatrix}{whether markers genes are already in preprocessed matrix form}

\item{genomen}{number of genes in the genome}

\item{metric}{adjusted p-value for hypergeometric test, or jaccard index}

\item{output_high}{if true (by default to fit with rest of package),
-log10 transform p-value}

\item{dr}{stored dimension reduction}

\item{seurat_out}{output cor matrix or called seurat object}

\item{set_ident}{whether the output seurat object has ident set to called type}

\item{...}{passed to matrixize_markers}
}
\description{
Main function to compare scRNA-seq data to gene lists.
}
