---
title: 'Cluster classification using mark gene lists'
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    vignette: >
      %\VignetteIndexEntry{clustifyR-genelists}
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

This vignette demonstrates how to apply ranked or unranked gene lists, based on prior knowledge of the cells/samples, to cluster classification. Here, average single cell RNA expression data for each cluster is used as input. The approach does work with full single cell expression matrix as well.

## Unranked list: Preprocessing gene list into dataframe/matrix form
An unranked gene list can be converted from Seurat::FindAllMarkers(), or assembled from vectors of genes, into a format as shown below:
```{r}
library(dplyr)
library(purrr)
library(stringr)
library(ComplexHeatmap)
```

```{r}
library(clustifyR)
# from dataframe generated by Seurat
pbmc4k_mm <- matrixize_markers(pbmc4k_markers)
head(pbmc4k_mm)

# or just lists of genes
NKm <- c("GNLY", "NKG7", "ZBTB32", "IL2RB")
DCm <- c("FCER1A", "CST3", "ITGAX", "IRF8")
Bm <- c("MS4A1", "CD79A","CD79B","CD37")
mm2 <- data.frame(Bm, DCm, NKm)
```

## Unranked list: Binarize single cell gene expression data
Cluster average gene expression is transformed into binary 0 or 1s (defaults to keeping the top1000 genes by expression)
```{r}
pbmc4k_avgb <- binarize_expr(pbmc4k_avg)
```

## Unranked list: Classify
Applying the gene lists to binary expression data ouputs a matrix of -log10 of adjusted P values for hypergeometric test, or jaccard index. Correct cluster identity is called when only 4 marker genes is provided in PBMCs.
```{r}
res <- compare_lists(pbmc4k_avgb,pbmc4k_mm)
Heatmap(res, cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))

res2 <- compare_lists(pbmc4k_avgb,mm2)
Heatmap(res2, cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))
```


## Ranked list: Preprocessing gene list into dataframe/matrix form
Arbitrary pseudo expression numbers are assigned when a ranked gene list is provided.
```{r}
pbmc4k_rm <- matrixize_markers(pbmc4k_markers, TRUE)
```

## Ranked list: Classify
Spearman correlation or other approaches implemented in the package can be applied.
```{r}
gene_constraints <- list(rownames(pbmc4k_avg),rownames(pbmc4k_rm))
pbmc4k_avg_d<- select_gene_subset(pbmc4k_avg, gene_constraints)
pbmc4k_rm_d <- select_gene_subset(pbmc4k_rm, gene_constraints)
out <- lapply(colnames(pbmc4k_avg_d),
              function(x){
                per_col <- lapply(colnames(pbmc4k_rm_d),
                                  function(y){
                                    compute_similarity(pbmc4k_avg_d[,x],
                                                       pbmc4k_rm_d[,y], corr_coef,
                                                       method = "spearman")})
                do.call(cbind, per_col)
              })

res <- do.call(rbind, out)
rownames(res) <- colnames(pbmc4k_avg_d)
colnames(res) <- colnames(pbmc4k_rm_d)

plot_cor(res,
         pbmc4k_meta,
         colnames(res)[1:7])
```

## Pancreas data from Indrop

Another example using pancreas single RNA seq data from Indrop method.

```{r, eval = FALSE}
# from dataframe generated by Seurat via SCE objects
betam <- c("INS", "IGF2", "IAPP","MAFA","NPTX2")
alpham <- c("GCG", "PDK4","LOXL2","IRX2","GC")
deltam <- c("SST", "RBP4","HHEX","PCSK1","LEPR")
mm2 <- data.frame(alpham, betam, deltam)

b <- binarize_expr(pan_indrop_avg)
res <- compare_lists(b,mm2)
Heatmap(res, cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))

res2 <- compare_lists(b,mm2,metric = "jaccard")
Heatmap(res2, cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "jaccard index"))

# not very good on per cell basis
# b2 <- binarize_expr(pan_indrop_matrix)
# res2 <- compare_lists(b2, mm2)
# plot_cor(res2,
#          pan_indrop_meta,
#          colnames(res2)[1:7])
```
