---
title: 'Cluster classification using mark gene lists'
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 3
    vignette: >
      %\VignetteIndexEntry{clustifyR-get_data}
      %\VignetteEngine{knitr::rmarkdown}
      %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

## Preprocessing gene list into dataframe/matrix form

```{r}
library('recount')
library('biomaRt')
library('dplyr')
library('purrr')
library('stringr')
library("ComplexHeatmap")
```

```{r}
library(clustifyR)
# from dataframe generated by Seurat
pbmc4k_mm <- matrixize_markers(pbmc4k_markers)
head(pbmc4k_mm)

# or just lists of genes
NKm <- c("GNLY", "NKG7")
DCm <- c("FCER1A", "CST3")
Bm <- c("MS4A1", "CD79A")
mm2 <- data.frame(Bm, DCm, NKm)
```

## Binarize single cell gene expression data
```{r}
pbmc4k_avgb <- binarize_expr(pbmc4k_avg)
```

## Classify

```{r}
res <- hyperp(pbmc4k_avgb,pbmc4k_mm)
Heatmap(-log10(res), cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))

res2 <- hyperp(pbmc4k_avgb,mm2)
Heatmap(-log10(res2), cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "-log10(adjP)"))
```
