<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data • clustifyR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data">
<meta property="og:description" content="">
<meta property="og:image" content="http://rnabioco.github.io/clustifyR/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyR</a>
        <span class="label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/clustifyR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/compare_methods.html">Compare Methods</a>
    </li>
    <li>
      <a href="../articles/exploratory_pca.html">Exploratory PCA</a>
    </li>
    <li>
      <a href="../articles/gene_list.html">Cluster classification using mark gene lists</a>
    </li>
    <li>
      <a href="../articles/get_recount_data.html">Downloading data from recount</a>
    </li>
    <li>
      <a href="../articles/random_forest.html">randomForest model</a>
    </li>
    <li>
      <a href="../articles/refine_cluster.html">Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/clustifyR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Clustering Single Cell RNA-Seq Data Using Reference Bulk RNA-Seq Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/clustifyR/blob/master/vignettes/refine_cluster.Rmd"><code>vignettes/refine_cluster.Rmd</code></a></small>
      <div class="hidden name"><code>refine_cluster.Rmd</code></div>

    </div>

    
    
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p>Clustering is a crucial step in single cell RNA-seq (scRNA-seq) data analysis. Most popular methods examine the inherent structure of the scRNA-seq dataset, without making use of the biological information of underlying cell types. Consequently, optimal determination of the number of clusters and interpretation of each cluster becomes a challenging problem. Here, we implement a clustering function, <code>refine_clusters</code>, which accounts for both the proximity of single cell expression levels within a cluster and the estabilished cell type information by comparing the single cell data to the reference bulk RNA-seq data of pre-defined cell types.</p>
</div>
<div id="theory" class="section level2">
<h2 class="hasAnchor">
<a href="#theory" class="anchor"></a>Theory</h2>
<p>Let the expression profile of a single cell be <span class="math inline">\(\vec{x\ }\)</span> where each element refers to the expression level (counts/reads) of a gene. Assume the scRNA-seq data is to be clustered into <span class="math inline">\(M\)</span> clusters, and the first <span class="math inline">\(N\)</span> clusters have reference bulk RNA-seq data of pre-defined cells. Here, we require <span class="math inline">\(M\geq N\)</span>. Let <span class="math inline">\(\vec{B_i}\)</span> be the expression profile of the bulk data for <span class="math inline">\(i\)</span>-th cluster, and we require <span class="math inline">\(\vec{x\ }\)</span> and <span class="math inline">\(\vec{B_i}\)</span> be organized in the same gene order.</p>
<p>The function <em>refine_clusters</em> defines the probability of a cell <span class="math inline">\(\vec{x\ }\)</span> being assigned to cluster <span class="math inline">\(i\)</span> as a weighted product of the probability that <span class="math inline">\(\vec{x\ }\)</span> is similar to <span class="math inline">\(\vec{B_i}\)</span> and the probability that <span class="math inline">\(\vec{x\ }\)</span> is proximal to the centroid of clsuter <span class="math inline">\(i\)</span> (with mean <span class="math inline">\(\vec{\mu_i}\)</span> and standard deviation <span class="math inline">\(\vec{\sigma_i}\)</span>).</p>
<p>The former probability is formulated as <span class="math display">\[P(\vec{x\ }; \vec{B_i}) \propto \exp\{\epsilon S(\vec{x\ }, \vec{B_i})\}\]</span> where <span class="math inline">\(S(\vec{x\ }, \vec{B_i})\)</span> represents the similarity score between <span class="math inline">\(\vec{x\ }\)</span> and <span class="math inline">\(\vec{B_i}\)</span>. One example of similarity score is the correlation computed by <em>run_cor</em> function. For clusters without reference data, we assign a default similarity score. <span class="math inline">\(\epsilon\)</span> is a user-defined weighting parameter.</p>
<p>The latter probability is modeled as a Gaussian distribution, i.e. <span class="math display">\[P(\vec{x\ }; \vec{\mu_i}, \vec{\sigma_i}) \sim N(\vec{\mu_i}, \mathbf{\Sigma}^{(i)})\]</span> where the covariance matrix <span class="math inline">\(\mathbf{\Sigma}^{(i)}\)</span> is a diagonal matrix where each element diagonal represents the variance of a gene’s expression within the cluster <span class="math inline">\(i\)</span> (<span class="math inline">\(\Sigma^{(i)}_{j,j} = \sigma_{i,j}^2\)</span>).</p>
<p>The overall probability is then expressed as <span class="math display">\[P(\vec{x\ }; \mathrm{Cluster}\ i) = P(\vec{x\ }; \vec{B_i})^\lambda P(\vec{x\ }; \vec{\mu_i}, \vec{\sigma_i})^{1-\lambda}\]</span> where <span class="math inline">\(\lambda\)</span> is a user-defined input parameter.</p>
<p>For each single cell <span class="math inline">\(\vec{x\ }\)</span>, the function <em>refine_clusters</em> computes the probability of it being assigned to each cluster, and assigns the cell to the cluster with highest probability. After assigning all single cells, a new clustering is formed and the function repeats the whole process until assignment is converged.</p>
<p>The algorithm works as follows</p>
<ul>
<li>Input: scRNA-seq expression matrix, bulk RNA-seq expression matrix, initial cluster assignment</li>
<li>Output: Optimal cluster assignment</li>
</ul>
<ol style="list-style-type: decimal">
<li>For each single cell (<span class="math inline">\(\vec{x\ }\)</span>) and each bulk RNA-seq data (<span class="math inline">\(\vec{B_i}\)</span>), compute the similarity score <span class="math inline">\(S(\vec{x\ }, \vec{B_i})\)</span>. For clusters without reference bulk data, a default similarity score is used.</li>
<li>For each cluster <span class="math inline">\(i\)</span>, compute its centroid <span class="math inline">\(\vec{\mu_i}\)</span> and <span class="math inline">\(\vec{\sigma_i}\)</span>.</li>
<li>For each cell (<span class="math inline">\(\vec{x\ }\)</span>) and each cluster <span class="math inline">\(i\)</span>, compute the probability of assignment <span class="math inline">\(P(\vec{x\ }; \mathrm{Cluster}\ i)\)</span>, and assign the cell to the cluster with highest probability.</li>
<li>Repeat Step 2-3 until convergence.</li>
</ol>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<div id="dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#dataset" class="anchor"></a>Dataset</h3>
<p>We make use of the <em>pbmc4k_matrix</em>, <em>pbmc4k_meta</em> and <em>pbmc_bulk_matrix</em> dataset. To satisfy the requirement of the function, we define six clusters (Monocytes, B Cells, T Cells, Dendritic Cells, NK Cells and Megakaryocytes) and re-assign each cell to one cluster based on its previous label. We further extract 5 samples (Monocytes, B Cells, T Cells, Dendritic Cells, NK Cells) from the bulk dataset to represent the reference expression profile for the first five clusters. No reference data for Megakaryocytes is provided.</p>
<pre class="preparation"><code>data("pbmc4k_meta"); data("pbmc4k_matrix"); data("pbmc_bulk_matrix"); data("pbmc4k_markers_M3Drop")

# construct sc_cluster
gene_constraints &lt;- get_common_elements(rownames(pbmc4k_matrix), 
                                         rownames(pbmc_bulk_matrix),
                                         pbmc4k_markers_M3Drop[,'Gene'])
sc_expr &lt;- pbmc4k_matrix[gene_constraints, ]
bulk_expr &lt;- pbmc_bulk_matrix[gene_constraints, ]
sc_tsne_coord &lt;- pbmc4k_meta[,c('tSNE_1', 'tSNE_2')];
sc_cluster &lt;- rep(0, ncol(sc_expr));
sc_cluster[which(pbmc4k_meta[,'classified']=="CD14+ Monocytes")] &lt;- 1; # 1 monocytes
sc_cluster[which(pbmc4k_meta[,'classified']=="FCGR3A+ Monocytes")] &lt;- 1;
sc_cluster[which(pbmc4k_meta[,'classified']=="B cells")] &lt;- 2; # 2 B cells
sc_cluster[which(pbmc4k_meta[,'classified']=="CD4 T cells, 1")] &lt;- 3; # 3 T cells
sc_cluster[which(pbmc4k_meta[,'classified']=="CD4 T cells, 2")] &lt;- 3;
sc_cluster[which(pbmc4k_meta[,'classified']=="CD8 T cells")] &lt;- 3;
sc_cluster[which(pbmc4k_meta[,'classified']=="Dendritic cells, 1")] &lt;- 4; # 4 DC
sc_cluster[which(pbmc4k_meta[,'classified']=="Dendritic Cells, 2?")] &lt;- 4;
sc_cluster[which(pbmc4k_meta[,'classified']=="NK cells")] &lt;- 5; # 5 NK cells
sc_cluster[which(pbmc4k_meta[,'classified']=="Megakaryocytes")] &lt;- 6; # 6 megakaryocytes (not found)
bulk_expr &lt;- bulk_expr[, c(3, 1, 7, 2, 5)]
colnames(bulk_expr) &lt;- c("Monocytes", "B Cells", "T Cells", "Dendritic Cells", "NK Cells");
cluster_names &lt;- c("Monocytes", "B Cells", "T Cells", "Dendritic Cells", "NK Cells", "Megakaryocytes");</code></pre>
</div>
<div id="execution" class="section level3">
<h3 class="hasAnchor">
<a href="#execution" class="anchor"></a>Execution</h3>
<p>To demonstrate the power of <code>refine_clusters</code>, we randomize the cluster assignment and use the function to re-assign clusters.</p>
<pre class="execution"><code>res &lt;- refine_clusters(sc_expr, sample(sc_cluster, length(sc_cluster)), bulk_expr, lambda=0.5, epsilon=1, if_compute_sigma=TRUE, num_iteration=20, default_similiarity=-1, disagreement_freq=0.01, if_plot=TRUE, tsne_coord=sc_tsne_coord, cluster_names=cluster_names, compute_method=corr_coef);</code></pre>
<p>For this execution, we put equal weights on the two probabilities and we allow the algorithm to run at most 20 iterations. Clustering is considered converged if less than 1% of cells are assigned with new labels during one iteration.</p>
<p>To evaluate the performance of the algorithm, we visualize the clusters on the t-SNE plot. As a reference, the ground truth is <img src="../inst/images/refine_cluster/initial_truth.png" alt="Ground Truth"></p>
<p>Before clustering, since assignment is randomized, the clusters are well-mixed on the t-SNE plot. <img src="../inst/images/refine_cluster/initial_random.png" alt="Before clustering"></p>
<p>The function shows rapid convergence after 1 iteration, <img src="../inst/images/refine_cluster/iter1.png" alt="After 1 iteration"></p>
<p>and clustering is stabilized after 10 iterations. <img src="../inst/images/refine_cluster/iter10.png" alt="After 10 iterations"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#motivation">Motivation</a></li>
      <li><a href="#theory">Theory</a></li>
      <li><a href="#example">Example</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Kent Riemondy, Austin Gillen, Rui Fu, Jay Hesselberth, Yue Hao, Chengzhe Tian, Michelle Daya.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
