<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compare scRNA-seq data to bulk RNA-seq data. — clustify • clustifyr</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Compare scRNA-seq data to bulk RNA-seq data. — clustify" />

<meta property="og:description" content="Compare scRNA-seq data to bulk RNA-seq data." />

<meta property="og:image" content="http://rnabioco.github.io/clustifyr/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">clustifyr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/benchmark.html">Benchmarking clustifyr</a>
    </li>
    <li>
      <a href="../articles/clustifyR.html">Introduction to clustifyr</a>
    </li>
    <li>
      <a href="../articles/otherformats.html">Support for various single cell object formats</a>
    </li>
    <li>
      <a href="../articles/othertools.html">Other considerations w/ clustifyr</a>
    </li>
    <li>
      <a href="../articles/references.html">Reference building for clustifyr</a>
    </li>
    <li>
      <a href="../articles/visualizations.html">Visualization of clustifyr results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rnabioco/clustifyr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare scRNA-seq data to bulk RNA-seq data.</h1>
    <small class="dont-index">Source: <a href='https://github.com/rnabioco/clustifyr/blob/master/R/main.R'><code>R/main.R</code></a></small>
    <div class="hidden name"><code>clustify.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Compare scRNA-seq data to bulk RNA-seq data.</p>
    
    </div>

    <pre class="usage"><span class='fu'>clustify</span>(<span class='no'>input</span>, <span class='no'>...</span>)

<span class='co'># S3 method for default</span>
<span class='fu'>clustify</span>(<span class='no'>input</span>, <span class='no'>ref_mat</span>, <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>n_perm</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>lookuptable</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rm0</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>obj_out</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='no'>...</span>)

<span class='co'># S3 method for seurat</span>
<span class='fu'>clustify</span>(<span class='no'>input</span>, <span class='no'>ref_mat</span>, <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>n_perm</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>, <span class='kw'>use_var_genes</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>rm0</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)

<span class='co'># S3 method for Seurat</span>
<span class='fu'>clustify</span>(<span class='no'>input</span>, <span class='no'>ref_mat</span>, <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>n_perm</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"spearman"</span>, <span class='kw'>use_var_genes</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"umap"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>rm0</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>rename_prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>single-cell expression matrix or Seurat object</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments to pass to compute_method function</p></td>
    </tr>
    <tr>
      <th>ref_mat</th>
      <td><p>reference expression matrix</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>cell cluster assignments, supplied as a vector or data.frame.
If data.frame is supplied then <code>cluster_col</code> needs to be set. Not required
if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata that contains cluster ids per cell.
Will default to first column of metadata if not supplied. Not required if
running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector of genes of interest to compare. If NULL, then
common genes between the expr_mat and ref_mat will be used for comparision.</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>if true run per cell, otherwise per cluster.</p></td>
    </tr>
    <tr>
      <th>n_perm</th>
      <td><p>number of permutations, set to 0 by default</p></td>
    </tr>
    <tr>
      <th>compute_method</th>
      <td><p>method(s) for computing similarity scores</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>whether to report certain variables chosen</p></td>
    </tr>
    <tr>
      <th>lookuptable</th>
      <td><p>if not supplied, will look in built-in table for object parsing</p></td>
    </tr>
    <tr>
      <th>rm0</th>
      <td><p>consider 0 as missing data, recommended for per_cell</p></td>
    </tr>
    <tr>
      <th>obj_out</th>
      <td><p>whether to output object instead of cor matrix</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum correlation score threshold, only used when obj_out = T</p></td>
    </tr>
    <tr>
      <th>use_var_genes</th>
      <td><p>if providing a seurat object, use the variable genes
(stored in seurat_object@var.genes) as the query_genes.</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>ref_mat</th>
      <td><p>reference expression matrix</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata that contains cluster ids per cell. Will default to first
column of metadata if not supplied. Not required if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector of genes of interest to compare. If NULL, then common genes between
the expr_mat and ref_mat will be used for comparision.</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>if true run per cell, otherwise per cluster.</p></td>
    </tr>
    <tr>
      <th>n_perm</th>
      <td><p>number of permutations, set to 0 by default</p></td>
    </tr>
    <tr>
      <th>compute_method</th>
      <td><p>method(s) for computing similarity scores</p></td>
    </tr>
    <tr>
      <th>use_var_genes</th>
      <td><p>if providing a seurat object, use the variable genes
(stored in seurat_object@var.genes) as the query_genes.</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum correlation score threshold</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>whether to report certain variables chosen</p></td>
    </tr>
    <tr>
      <th>rm0</th>
      <td><p>consider 0 as missing data, recommended for per_cell</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments to pass to compute_method function</p></td>
    </tr>
    <tr>
      <th>ref_mat</th>
      <td><p>reference expression matrix</p></td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>column in metadata that contains cluster ids per cell. Will default to first
column of metadata if not supplied. Not required if running correlation per cell.</p></td>
    </tr>
    <tr>
      <th>query_genes</th>
      <td><p>A vector of genes of interest to compare. If NULL, then common genes between
the expr_mat and ref_mat will be used for comparision.</p></td>
    </tr>
    <tr>
      <th>per_cell</th>
      <td><p>if true run per cell, otherwise per cluster.</p></td>
    </tr>
    <tr>
      <th>n_perm</th>
      <td><p>number of permutations, set to 0 by default</p></td>
    </tr>
    <tr>
      <th>compute_method</th>
      <td><p>method(s) for computing similarity scores</p></td>
    </tr>
    <tr>
      <th>use_var_genes</th>
      <td><p>if providing a seurat object, use the variable genes
(stored in seurat_object@var.genes) as the query_genes.</p></td>
    </tr>
    <tr>
      <th>dr</th>
      <td><p>stored dimension reduction</p></td>
    </tr>
    <tr>
      <th>seurat_out</th>
      <td><p>output cor matrix or called seurat object</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>identity calling minimum correlation score threshold</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>whether to report certain variables chosen</p></td>
    </tr>
    <tr>
      <th>rm0</th>
      <td><p>consider 0 as missing data, recommended for per_cell</p></td>
    </tr>
    <tr>
      <th>rename_prefix</th>
      <td><p>prefix to add to type and r column names</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments to pass to compute_method function</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>matrix of correlation values, clusters from input as row names, cell
types from ref_mat as column names</p>
<p>seurat2 object with type assigned in metadata, or matrix of
correlation values, clusters from input as row names, cell types from
ref_mat as column names</p>
<p>seurat3 object with type assigned in metadata, or matrix of
correlation values, clusters from input as row names, cell types from
ref_mat as column names</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Annotate a matrix and metadata</span>
<span class='fu'>clustify</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>pbmc_matrix_small</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>pbmc_meta</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>pbmc_bulk_matrix</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='no'>pbmc_vargenes</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"classified"</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='output co'>#&gt; <span class='message'>using # of genes: 1688</span></div><div class='output co'>#&gt;              primary human B cells rep 1 primary human myeloid DC rep 1
#&gt; B                             0.62161187                      0.5134249
#&gt; CD14+ Mono                    0.38176061                      0.6585926
#&gt; CD8 T                         0.43926595                      0.4648890
#&gt; DC                            0.37326478                      0.5726146
#&gt; FCGR3A+ Mono                  0.39707564                      0.6828512
#&gt; Memory CD4 T                  0.47496679                      0.4835943
#&gt; Naive CD4 T                   0.47847311                      0.4753289
#&gt; NK                            0.41142231                      0.4600473
#&gt; Platelet                      0.09448371                      0.2607297
#&gt;              primary human monocytes rep 1 primary human neutrophils rep 1
#&gt; B                                0.4858326                       0.4004996
#&gt; CD14+ Mono                       0.6932047                       0.5548357
#&gt; CD8 T                            0.4396090                       0.3759313
#&gt; DC                               0.5619668                       0.3990601
#&gt; FCGR3A+ Mono                     0.6464808                       0.5302247
#&gt; Memory CD4 T                     0.4637761                       0.3883083
#&gt; Naive CD4 T                      0.4615861                       0.3824808
#&gt; NK                               0.4339593                       0.3824971
#&gt; Platelet                         0.2607125                       0.2120760
#&gt;              primary human NK cells rep 1 primary human PBMC rep 1
#&gt; B                               0.4682206                0.4930223
#&gt; CD14+ Mono                      0.4380704                0.5716609
#&gt; CD8 T                           0.5988312                0.5230516
#&gt; DC                              0.4031306                0.4926074
#&gt; FCGR3A+ Mono                    0.4434886                0.5511974
#&gt; Memory CD4 T                    0.5500075                0.5277524
#&gt; Naive CD4 T                     0.5418622                0.5293089
#&gt; NK                              0.6443927                0.5141864
#&gt; Platelet                        0.1429789                0.2285756
#&gt;              primary human T cells rep 1 primary human B cells rep 2
#&gt; B                              0.4688555                   0.6409642
#&gt; CD14+ Mono                     0.3525068                   0.4081323
#&gt; CD8 T                          0.5533326                   0.4529453
#&gt; DC                             0.3248138                   0.4027535
#&gt; FCGR3A+ Mono                   0.3585348                   0.4240105
#&gt; Memory CD4 T                   0.6017151                   0.4884443
#&gt; Naive CD4 T                    0.6091448                   0.4899223
#&gt; NK                             0.5180162                   0.4274134
#&gt; Platelet                       0.0734803                   0.1147430
#&gt;              primary human myeloid DC rep 2 primary human monocytes rep 2
#&gt; B                                 0.5142264                     0.4817698
#&gt; CD14+ Mono                        0.6585599                     0.7061875
#&gt; CD8 T                             0.4657318                     0.4380312
#&gt; DC                                0.5714870                     0.5624386
#&gt; FCGR3A+ Mono                      0.6868239                     0.6588797
#&gt; Memory CD4 T                      0.4841326                     0.4637714
#&gt; Naive CD4 T                       0.4743983                     0.4603821
#&gt; NK                                0.4652355                     0.4398213
#&gt; Platelet                          0.2626675                     0.2685089
#&gt;              primary human neutrophils rep 2 primary human NK cells rep 2
#&gt; B                                  0.3882407                    0.4719677
#&gt; CD14+ Mono                         0.5573050                    0.4408221
#&gt; CD8 T                              0.3636086                    0.6007341
#&gt; DC                                 0.3955419                    0.4074878
#&gt; FCGR3A+ Mono                       0.5322509                    0.4450197
#&gt; Memory CD4 T                       0.3760267                    0.5502540
#&gt; Naive CD4 T                        0.3692816                    0.5415756
#&gt; NK                                 0.3715250                    0.6474230
#&gt; Platelet                           0.2165613                    0.1468565
#&gt;              primary human PBMC rep 2 primary human T cells rep 2
#&gt; B                           0.5179320                  0.48219201
#&gt; CD14+ Mono                  0.6026460                  0.36879949
#&gt; CD8 T                       0.5235911                  0.57153550
#&gt; DC                          0.5077886                  0.33963841
#&gt; FCGR3A+ Mono                0.5802021                  0.37464917
#&gt; Memory CD4 T                0.5471137                  0.61932481
#&gt; Naive CD4 T                 0.5498952                  0.62715073
#&gt; NK                          0.5070531                  0.53661318
#&gt; Platelet                    0.2523581                  0.08479795</div><div class='input'>
<span class='co'># Annotate using a different method</span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>clustify</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>pbmc_matrix_small</span>,
  <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>pbmc_meta</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>pbmc_bulk_matrix</span>,
  <span class='kw'>query_genes</span> <span class='kw'>=</span> <span class='no'>pbmc_vargenes</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"classified"</span>,
  <span class='kw'>compute_method</span> <span class='kw'>=</span> <span class='st'>"cosine"</span>
)

<span class='co'># Annotate a Seurat object</span>
<span class='fu'>clustify</span>(
  <span class='no'>s_small</span>,
  <span class='no'>pbmc_bulk_matrix</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"res.1"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"tsne"</span>
)</div><div class='output co'>#&gt; <span class='message'>seurat not loaded, returning cor_mat instead</span></div><div class='output co'>#&gt;   primary human B cells rep 1 primary human myeloid DC rep 1
#&gt; 0                  0.56495413                      0.2905662
#&gt; 1                  0.38339099                      0.6645159
#&gt; 2                  0.01569994                      0.3132044
#&gt; 3                 -0.42255352                     -0.1581610
#&gt;   primary human monocytes rep 1 primary human neutrophils rep 1
#&gt; 0                     0.2860993                      0.31382849
#&gt; 1                     0.5913028                      0.54143473
#&gt; 2                     0.2506601                      0.26686921
#&gt; 3                    -0.1596368                     -0.04526416
#&gt;   primary human NK cells rep 1 primary human PBMC rep 1
#&gt; 0                    0.2521516                0.4810757
#&gt; 1                    0.3960676                0.4553994
#&gt; 2                    0.5476851                0.3873399
#&gt; 3                   -0.2740145               -0.2484332
#&gt;   primary human T cells rep 1 primary human B cells rep 2
#&gt; 0                   0.5081027                   0.5427781
#&gt; 1                   0.4296189                   0.3996944
#&gt; 2                   0.3657692                   0.0329662
#&gt; 3                  -0.3359033                  -0.3573163
#&gt;   primary human myeloid DC rep 2 primary human monocytes rep 2
#&gt; 0                      0.2512582                    0.21316182
#&gt; 1                      0.6595903                    0.57611050
#&gt; 2                      0.2938857                    0.22226395
#&gt; 3                     -0.1374992                   -0.09252722
#&gt;   primary human neutrophils rep 2 primary human NK cells rep 2
#&gt; 0                      0.29267862                    0.3189304
#&gt; 1                      0.55990915                    0.4737587
#&gt; 2                      0.21612700                    0.5310227
#&gt; 3                     -0.03016533                   -0.1999765
#&gt;   primary human PBMC rep 2 primary human T cells rep 2
#&gt; 0                0.4634318                   0.5432248
#&gt; 1                0.4997303                   0.4406714
#&gt; 2                0.2883316                   0.3448771
#&gt; 3               -0.1891536                  -0.2745370</div><div class='input'>
<span class='co'># Annotate (and return) a Seurat object per-cell</span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>clustify</span>(
  <span class='kw'>input</span> <span class='kw'>=</span> <span class='no'>s_small</span>,
  <span class='kw'>ref_mat</span> <span class='kw'>=</span> <span class='no'>pbmc_bulk_matrix</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"res.1"</span>,
  <span class='kw'>seurat_out</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>per_cell</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>dr</span> <span class='kw'>=</span> <span class='st'>"tsne"</span>
)</div><div class='output co'>#&gt; <span class='warning'>Warning: the standard deviation is zero</span></div><div class='output co'>#&gt; <span class='message'>seurat not loaded, returning cor_mat instead</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rui Fu, Kent Riemondy, Austin Gillen, Chengzhe Tian, Jay Hesselberth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


